File: app/src/main/java/com/example/mobiledevelopmentassignment2/ui/MainActivity.java
package com.example.mobiledevelopmentassignment2.ui;

import com.example.mobiledevelopmentassignment2.R;
import android.annotation.SuppressLint;
import android.database.Cursor;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;


public class MainActivity extends AppCompatActivity {

    private DatabaseHelper dbHelper;
    private EditText addressInput, latitudeInput, longitudeInput, idInput;
    private TextView resultView;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize DatabaseHelper
        dbHelper = new DatabaseHelper(this);

        // Initialize UI components
        addressInput = findViewById(R.id.addressInput);
        latitudeInput = findViewById(R.id.latitudeInput);
        longitudeInput = findViewById(R.id.longitudeInput);
        resultView = findViewById(R.id.resultView);

        // Set up button listeners
        findViewById(R.id.queryButton).setOnClickListener(v -> queryLocation());
        findViewById(R.id.addButton).setOnClickListener(v -> addLocation());
        findViewById(R.id.updateButton).setOnClickListener(v -> updateLocation());
        findViewById(R.id.deleteButton).setOnClickListener(v -> deleteLocation());
    }

    @SuppressLint("SetTextI18n")
    private void queryLocation() {
        String address = addressInput.getText().toString();
        Cursor cursor = dbHelper.getLocation(address);
        if (cursor != null && cursor.moveToFirst()) {
            double latitude = cursor.getDouble(cursor.getColumnIndexOrThrow("latitude"));
            double longitude = cursor.getDouble(cursor.getColumnIndexOrThrow("longitude"));
            resultView.setText("Latitude: " + latitude + ", Longitude: " + longitude);
            cursor.close();
        } else {
            resultView.setText("Address not found.");
        }
    }

    @SuppressLint("SetTextI18n")
    private void addLocation() {
        String address = addressInput.getText().toString();
        double latitude = Double.parseDouble(latitudeInput.getText().toString());
        double longitude = Double.parseDouble(longitudeInput.getText().toString());
        dbHelper.addLocation(address, latitude, longitude);
        resultView.setText("Location added.");
    }

    @SuppressLint("SetTextI18n")
    private void updateLocation() {
        int id = Integer.parseInt(idInput.getText().toString());
        String address = addressInput.getText().toString();
        double latitude = Double.parseDouble(latitudeInput.getText().toString());
        double longitude = Double.parseDouble(longitudeInput.getText().toString());
        dbHelper.updateLocation(id, address, latitude, longitude);
        resultView.setText("Location updated.");
    }

    @SuppressLint("SetTextI18n")
    private void deleteLocation() {
        int id = Integer.parseInt(idInput.getText().toString());
        dbHelper.deleteLocation(id);
        resultView.setText("Location deleted.");
    }
}

File: app/src/main/java/com/example/mobiledevelopmentassignment2/MainActivity.kt
package com.example.mobiledevelopmentassignment2

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.example.mobiledevelopmentassignment2.ui.theme.MobileDevelopmentAssignment2Theme

class MainActivity : ComponentActivity() {
    private lateinit var dbHelper: DataBaseHelper

    @SuppressLint("UnusedMaterial3ScaffoldPaddingParameter")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dbHelper = DataBaseHelper(this)

        setContent {
            MobileDevelopmentAssignment2Theme {
                Scaffold(modifier = Modifier.fillMaxSize()) {
                    LocationFinderApp(dbHelper)
                }
            }
        }
    }
}

@Composable
fun LocationFinderApp(dbHelper: DataBaseHelper) {
    var address by remember { mutableStateOf("") }
    var latitude by remember { mutableStateOf("") }
    var longitude by remember { mutableStateOf("") }
    var result by remember { mutableStateOf("Result will appear here.") }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        // Address Input
        OutlinedTextField(
            value = address,
            onValueChange = { address = it },
            label = { Text("Address") },
            modifier = Modifier.fillMaxWidth()
        )

        // Latitude Input
        OutlinedTextField(
            value = latitude,
            onValueChange = { latitude = it },
            label = { Text("Latitude") },
            modifier = Modifier.fillMaxWidth()
        )

        // Longitude Input
        OutlinedTextField(
            value = longitude,
            onValueChange = { longitude = it },
            label = { Text("Longitude") },
            modifier = Modifier.fillMaxWidth()
        )

        Spacer(modifier = Modifier.height(16.dp))

        // Query Button
        Button(
            onClick = {
                val cursor = dbHelper.getLocation(address)
                if (cursor != null && cursor.moveToFirst()) {
                    val lat = cursor.getDouble(0)
                    val lon = cursor.getDouble(1)
                    result = "Latitude: $lat, Longitude: $lon"
                    cursor.close()
                } else {
                    result = "Address not found."
                }
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Query Location")
        }

        // Add Button
        Button(
            onClick = {
                val lat = latitude.toDoubleOrNull() ?: 0.0
                val lon = longitude.toDoubleOrNull() ?: 0.0
                dbHelper.addLocation(address, lat, lon)
                result = "Location added."
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Add Location")
        }

        // Update Button
        Button(
            onClick = {
                val lat = latitude.toDoubleOrNull() ?: 0.0
                val lon = longitude.toDoubleOrNull() ?: 0.0
                dbHelper.updateLocation(address, lat, lon)
                result = "Location updated."
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Update Location")
        }

        // Delete Button
        Button(
            onClick = {
                dbHelper.deleteLocation(address)
                result = "Location deleted."
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Delete Location")
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Display Result
        Text(text = result, modifier = Modifier.padding(8.dp))
    }
}

@Preview(showBackground = true)
@Composable
fun PreviewLocationFinderApp() {
    MobileDevelopmentAssignment2Theme {
        LocationFinderApp(DataBaseHelper(null))  // Pass a dummy DatabaseHelper for preview purposes
    }
}
